doc/python-grader/index.md

```diff
@@ -0,0 +1,163 @@
+
+  <p>The setup code gives the following variables:</p>
+  <p><pl-variable-description variables-category="names_for_user"></pl-variable-description></p>
+
+  <p>Your code snippet should define the following variables:</p>
+  <pl-variable-description variables-category="names_from_user"></pl-variable-description>
+  <pl-file-editor
+    file_name="user_code.py"
+    ace_mode="ace/mode/python"
+    source-file-name="tests/initial_code.py"></pl-file-editor>
+</pl-question-panel>
+
+<pl-submission-panel>
+  <pl-external-grader-results />
+</pl-submission-panel>
+```
+
+### `tests/setup_code.py`
+
+This file is executed before any reference or student code is run.  Any variables defined in `names_for_user` can be accessed from here in student code, while the reference answer may freely access variables without restriction.
+
+### `tests/test.py`
+
+The test cases for each coding problem are defined as methods of a `test` class contained in the aptly named `test.py` file.  The class will have one of the following signatures, depending if you require plots from the student:
+
+```python
+## No plot grading
+class Test(PrairieLearnTestCase):
+
+## Plot grading enabled
+class Test(PrairieLearnTestCaseWithPlot):
+```
+
+Each test case is a separate method in the class, the names of these functions must be prefixed with `test_`, but any name that follows is arbitrary.  Test cases are run in sorted order by the `unittest` library, so one convention that can be used is to give tests numeric names to ensure their running order.
+
+Adding a name and point value to the test case is done by means of python decorators:
+`@points(val)` and `@name(name of the test case)`.  These will control the name of the case and the points awarded in the "Test Results" dropdown menu that are shown when a student submits a solution.  An example of a definition:
+
+```python
+@points(10)
+@name("Cool test case")
+def test_0(self):
+```
+
+Inside the test case implementation, the student answer variables and reference answer variables can be accessed as children of the tuples `self.st` and `self.ref`, respectively.  There are various helper functions to check correctness of different types of variables, these are defined in `code_feedback.py`.  These are taken from the RELATE grader, so this may be familiar to those with prior experience with RELATE.
+
+At the end of the test case, set the correctness of the answer using `feedback.set_points()`.  This function takes a floating point number between 0 and 1 (inclusive), with 0 being completely *in*correct and 1 being completely correct.  By default, if no points are given the test case will be marked incorrect.
+
+The overall structure of a test case should look something like:
+
+```python
+@points(10)
+@name("name of the test case")
+def test_0(self):
+   if feedback.check_scalar('name of the variable', self.ref.variable_name, self.st.variable_names):
+       Feedback.set_points(1)
+	else:
+		Feedback.set_points(0)
+``` 
+
```

Comment:
Maybe add a note that if `set_points(.5)` this means that 50% of points from the `@points(10)` directive will be awarded?

Reply:
Yes, and I’m in favour of even renaming the `set_points()` to something more clear like `set_percent()` even though this might break existing code (sorry, cs357)

Reply:
@nicknytko yeah, I think that would be better. Defer to @mwest1066 on that.